From 73d919454db86600b0799f3d7cada7b8b44fe298 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:03:57 -0400
Subject: [powerpc] Don't flush/invalidate the d/icache for an unknown relocation type

Message-id: <1410545655-205645-9-git-send-email-dzickus@redhat.com>
Patchwork-id: 94063
O-Subject: [RHEL7 PATCH 008/626] powerpc: Don't flush/invalidate the d/icache for an unknown relocation type
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit 348c2298a6fd2b145e789739808d5e7598e275fc
Author: Kevin Hao <haokexin@gmail.com>
Date:   Thu Jun 27 09:09:43 2013 +0800

    powerpc: Don't flush/invalidate the d/icache for an unknown relocation type

    For an unknown relocation type since the value of r4 is just the 8bit
    relocation type, the sum of r4 and r7 may yield an invalid memory
    address. For example:
        In normal case:
                 r4 = c00xxxxx
                 r7 = 40000000
                 r4 + r7 = 000xxxxx

        For an unknown relocation type:
                 r4 = 000000xx
                 r7 = 40000000
                 r4 + r7 = 400000xx
       400000xx is an invalid memory address for a board which has just
       512M memory.

    And for operations such as dcbst or icbi may cause bus error for an
    invalid memory address on some platforms and then cause the board
    reset. So we should skip the flush/invalidate the d/icache for
    an unknown relocation type.

    Signed-off-by: Kevin Hao <haokexin@gmail.com>
    Acked-by: Suzuki K. Poulose <suzuki@in.ibm.com>
    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/kernel/reloc_32.S b/arch/powerpc/kernel/reloc_32.S
index ef46ba6..f366fed 100644
--- a/arch/powerpc/kernel/reloc_32.S
+++ b/arch/powerpc/kernel/reloc_32.S
@@ -166,7 +166,7 @@ ha16:
  /* R_PPC_ADDR16_LO */
 lo16:
  cmpwi r4, R_PPC_ADDR16_LO
- bne nxtrela
+ bne unknown_type
  lwz r4, 0(r9) /* r_offset */
  lwz r0, 8(r9) /* r_addend */
  add r0, r0, r3
@@ -191,6 +191,7 @@ nxtrela:
  dcbst r4,r7
  sync   /* Ensure the data is flushed before icbi */
  icbi r4,r7
+unknown_type:
  cmpwi r8, 0  /* relasz = 0 ? */
  ble done
  add r9, r9, r6 /* move to next entry in the .rela table */
-- 
1.7.1